### PR TITLE
upgrade UCSDK version

### PR DESCRIPTION
**Pull request by yiming-amzn**
*Issue #, if available:*

*Description of changes:*
* upgrade UCSDK version to fix missed call issues for Citrix VDI redirection
* Fixing race conditions between UCSDK init and env check to avoid webrtc failure error message during initialization
* 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/js/strategies/CitrixVDIStrategy.js [hunk 1]
```diff
@@ -6,28 +6,41 @@ import CCPInitiationStrategyInterface from "./CCPInitiationStrategyInterface";
 import {hitch} from "../utils";
 import {FailedState} from "../rtc_session";
 import {RTC_ERRORS} from "../rtc_const";
-import "@citrix/ucsdk/CitrixWebRTC";
+
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:9]: nit: extra line

File: src/js/strategies/CitrixVDIStrategy.js [hunk 2]
```diff
@@ -6,28 +6,41 @@ import CCPInitiationStrategyInterface from "./CCPInitiationStrategyInterface";
...
-            const registryValue = Promise.resolve(1);
-            return new Promise(function (resolve, reject) {
-                //retrieve registry value internally
-                registryValue.then((v) => {
-                    resolve(v);
-                }).catch(() => {
-                    reject();
-                })
-            })
+        if(useRealCitrix){
```
Comment (siwei-w) [inline src/js/strategies/CitrixVDIStrategy.js:16]: [Query] 
1. Does `useRealCitrix` intend to import Citrix WebRTC SDK dynamically?
2. After this change, will we see HDX related logs in the console for native CCP?

File: src/js/strategies/CitrixVDIStrategy.js [hunk 3]
```diff
@@ -6,28 +6,41 @@ import CCPInitiationStrategyInterface from "./CCPInitiationStrategyInterface";
...
+
+    initCitrixWebRTC() {
+        window.CitrixWebRTC.setVMEventCallback((event) => {
+            if (event.event === 'vdiClientConnected') {
+                if (!window.CitrixWebRTC.isFeatureOn("webrtc1.0")) {
+                    throw new Error('Citrix WebRTC redirection feature is NOT supported!');
+                }
+                console.log("CitrixVDIStrategy initialized");
+            } else if (event.event === 'vdiClientDisconnected') {
+                console.log("vdiClientDisconnected");
```
Comment (siwei-w) [inline src/js/strategies/CitrixVDIStrategy.js:33]: [Query] Do we need to throw exception if VDI client is disconnected?
Maybe we could log this using CCP logger as well.

File: src/js/strategies/CitrixVDIStrategy.js [hunk 4]
```diff
@@ -6,28 +6,41 @@ import CCPInitiationStrategyInterface from "./CCPInitiationStrategyInterface";
...
-                }).catch(() => {
-                    reject();
-                })
-            })
+        if(useRealCitrix){
+            require("@citrix/ucsdk/CitrixWebRTC");
         }
+        console.log("CitrixVDIStrategy initializing");
+        this.initCitrixWebRTC();
+        this.initGetCitrixWebrtcRedir();
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:21]: Should this be part of initCitrixWebRTC()?

File: src/js/strategies/CitrixVDIStrategy.js [hunk 5]
```diff
@@ -6,28 +6,41 @@ import CCPInitiationStrategyInterface from "./CCPInitiationStrategyInterface";
...
+        this.initLog();
+    }
+
+    initCitrixWebRTC() {
+        window.CitrixWebRTC.setVMEventCallback((event) => {
+            if (event.event === 'vdiClientConnected') {
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:27]: What other events do we get from Citrix? Would it logging other events be helpful in troubleshooting?

File: src/js/strategies/CitrixVDIStrategy.js (hunk 3 above)
Comment (Vignendra) [reply src/js/strategies/CitrixVDIStrategy.js:33]: There could be cases where the disconnection was done gracefully. So exception might not always be the right choice here?

File: src/js/strategies/CitrixVDIStrategy.js (hunk 4 above)
Comment (Vignendra) [reply src/js/strategies/CitrixVDIStrategy.js:21]: Any reason we cannot club all these separate functions in a single init function?  Seems like they are all related

File: test/unit/strategies/CitrixVDIStrategy.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        delete global.window;
+        delete global.connect;
+    });
+
+    describe('constructor', () => {
+        it('should initialize CitrixWebRTC, getCitrixWebrtcRedir, and logging', () => {
+            const instance = new CitrixVDIStrategy(false);
+
+            chai.expect(global.window.CitrixWebRTC.setVMEventCallback).to.have.been.calledOnce;
+            chai.expect(global.window.getCitrixWebrtcRedir).to.be.a('function');
```
Comment (Vignendra) [inline test/unit/strategies/CitrixVDIStrategy.js:30]: is it possible to have a stronger assertion?

File: package.json [hunk 7]
```diff
@@ -23,11 +23,11 @@
     "version": "grunt build && grunt copyForPublish && git add package.json release/* ",
     "postversion": "export GITTAG=\"echo $(git describe --abbrev=0 --tags | sed 's/^v//')\" && git push --force --set-upstream origin bumpVersion --follow-tags && git checkout gh-pages && git pull && cp out/connect-rtc-debug.js ./connect-rtc-debug-`$GITTAG`.js && cp out/connect-rtc.min.js ./connect-rtc-`$GITTAG`.min.js && ln -fs connect-rtc-debug-`$GITTAG`.js connect-rtc-debug-latest.js && ln -fs connect-rtc-`$GITTAG`.min.js connect-rtc-latest.min.js && git add connect-rtc*.js && git commit -m `$GITTAG` && git push --set-upstream origin gh-pages && git checkout master",
     "prepublish": "grunt build",
-    "test": "grunt && nyc --reporter=html mocha --compilers js:babel-core/register test/unit"
+    "test": "grunt && nyc --reporter=html mocha --compilers js:babel-core/register test/unit/**/*.js test/unit/*.js"
   },
   "_ucsdk_upgrade_note": "Please follow the instruction here(https://tiny.amazon.com/18mszf5px/quipaTMFStep) to upgrade @citrix/ucsdk",
   "dependencies": {
-    "@citrix/ucsdk": "2.0.5",
+    "@citrix/ucsdk": "3.1.0",
```
Comment (Vignendra) [inline package.json:30]: Does this require legal team's approval?

File: test/unit/strategies/CitrixVDIStrategy.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            );
+        });
+
+        it('should log "CitrixVDIStrategy initialized" if WebRTC redirection feature is supported', () => {
+            const instance = new CitrixVDIStrategy(false);
+            global.window.CitrixWebRTC.isFeatureOn.returns(true);
+
+            const callback = global.window.CitrixWebRTC.setVMEventCallback.getCall(0).args[0];
+            callback({ event: 'vdiClientConnected' });
+            chai.assert(console.log.calledWith('CitrixVDIStrategy initialized'));
```
Comment (Vignendra) [inline test/unit/strategies/CitrixVDIStrategy.js:52]: Can this assertion just be part of the first unit test where we are testing all successful paths? I think we can club a few of these unit tests together.

File: package.json (hunk 7 above)
Comment (yiming-amzn) [reply package.json:30]: nope, I verified that the legal term did not change between 2.0.5 and 3.1.0

File: src/js/strategies/CitrixVDIStrategy.js (hunk 2 above)
Comment (yiming-amzn) [reply src/js/strategies/CitrixVDIStrategy.js:14]: 1. Yes
2. with 3.1.0, HDX related log is gone as HDX init would not happen automatically, unless we can initUCSDK. However, based on the unit test behavior, I would assume v2.0.5 will not have HDX related log with this change

File: src/js/strategies/CitrixVDIStrategy.js (hunk 5 above)
Comment (yiming-amzn) [reply src/js/strategies/CitrixVDIStrategy.js:25]: when setup callback before initUCSDK, it is intended to be used for checking initialization success or failure status.
The rest of the fields are relevant to OS info that we have via userAgent in call end report, and citrix versino info. 

```
{ "event":"vdiClientConnected",
"type_script": { "type":5, "version":"1.6.0.1"},
"webrpc": {"type":0, "version":"1.6.0."},
"webrtc_codecs":{"type":1, "version":"1.76.0.0"},
"receiver": {"type":2, "version":"19.12.0.14"},
"vda": {"type":3, "version":"0.0.0.0"},
"endpoint": {"type":4, "version":"0.0.0.0"}
"osinfo": { "architecture":"x86_64",
"distro":"Microsoft",
"edition":"Windows 10 Enterprise\u0000",
"family":"Windows",
"version":"6.3\u0000"}
}
```

Comment (yiming-amzn) [reply src/js/strategies/CitrixVDIStrategy.js:25]: I think most of those version info is also printed in the Citrix log, and we collected it in CCP Log

File: test/unit/strategies/CitrixVDIStrategy.js (hunk 6 above)
Comment (yiming-amzn) [reply test/unit/strategies/CitrixVDIStrategy.js:30]: Citrix need this `window.getCitrixWebrtcRedir ` to be a function, with return value being a promise of value 0, 1, or empty

File: src/js/strategies/CitrixVDIStrategy.js (hunk 4 above)
Comment (yiming-amzn) [reply src/js/strategies/CitrixVDIStrategy.js:19]: no specific reason or preference, but trying to make it more readable

File: src/js/strategies/CitrixVDIStrategy.js (hunk 3 above)
Comment (yiming-amzn) [reply src/js/strategies/CitrixVDIStrategy.js:31]: based on the documentation, my understanding is this disconnected would happen only if the env setup is wrong, or agent disconnect from Citrix VDI. As we are not seeing any missed call or dropped call due to this issue, I assume we can leave this as a potential improvement item and see if we need to handle it based on the agent experience

File: test/unit/strategies/CitrixVDIStrategy.js (hunk 8 above)
Comment (yiming-amzn) [reply test/unit/strategies/CitrixVDIStrategy.js:52]: yep I think that works

File: src/js/strategies/CitrixVDIStrategy.js [hunk 9]
```diff
@@ -99,18 +97,20 @@ export default class CitrixVDIStrategy extends CCPInitiationStrategyInterface {
         return new window.CitrixWebRTC.CitrixPeerConnection(configuration, optionalConfiguration);
     }
 
-    connect(self) {
-        self._pc.onaddstream = hitch(self, self._ontrack);
```
[the remark this answers is not in the document]
Comment (yiming-amzn) [reply src/js/strategies/CitrixVDIStrategy.js:90]: Yep this is the biggest change of this PR, and I requested the Principle Engineer on the Citrix side to review this change. I just forwarded the signoff from their PE on the code snippet

[the remark this answers is not in the document]
Comment (yiming-amzn) [reply src/js/strategies/CitrixVDIStrategy.js:90]: Here are the manual test that I performed : https://quip-amazon.com/TSjbAJSK4nU6/Manual-Test-Cases-for-lily-hdx-rtcjs, including long duration calls and back to back calls for several hours

File: src/js/strategies/CitrixVDIStrategy.js [hunk 10]
```diff
@@ -99,18 +97,20 @@ export default class CitrixVDIStrategy extends CCPInitiationStrategyInterface {
...
     _ontrack(self, evt) {
-        const remoteStream = evt.stream.clone();
-
-        const audioTracks = evt.stream.getAudioTracks();
-        if (audioTracks !== undefined && audioTracks.length > 0) {
-            self._remoteAudioStream = remoteStream;
-            self._remoteAudioElement.srcObject = remoteStream;
+        window.CitrixWebRTC.mapAudioElement(self._remoteAudioElement);
+        if (evt.streams.length > 1) {
+            self._logger.warn('Found more than 1 streams for ' + evt.track.kind + ' track ' + evt.track.id + ' : ' +
```
Comment (Vignendra) [inline src/js/strategies/CitrixVDIStrategy.js:102]: https://developer.mozilla.org/en-US/docs/Web/API/Media_Capture_and_Streams_API.  Concepts of MediaStream and track are a little jumbled here. Based on offline discussion - This has been copied from existing strategy and that piece of code has been there for a long time. In order to make major changes, we want to keep this similar.

We can take a backlog AI to cleanup this code through minimal change iterations in the future.

**Review by Vignendra — APPROVED** [review]

**Review by siwei-w — APPROVED** [review]

**Pull request merged**